sqlite.go

```diff
@@ -0,0 +1,530 @@
+	_ = row // TODO: return error if exec on query which returns rows?
+	if err != nil {
+		return nil, err
+	}
+	return stmtResult{lastInsertID: lastInsertRowID, rowsAffected: changes}, nil
+}
+
+type stmtResult struct {
+	lastInsertID int64
+	rowsAffected int64
+}
+
+func (res stmtResult) LastInsertId() (int64, error) { return res.lastInsertID, nil }
+func (res stmtResult) RowsAffected() (int64, error) { return res.rowsAffected, nil }
+
+func (s *stmt) QueryContext(ctx context.Context, args []driver.NamedValue) (driver.Rows, error) {
+	if err := s.stmt.Reset(); err != nil {
+		return nil, s.reserr("Stmt.Query(Reset)", err)
+	}
+	if err := s.bindAll(args); err != nil {
+		return nil, err
+	}
+	return &rows{stmt: s}, nil
+}
+
+func (s *stmt) bindAll(args []driver.NamedValue) error {
+	for _, arg := range args {
+		if err := s.bind(arg); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *stmt) bind(arg driver.NamedValue) error {
+	var debugName interface{} = arg.Ordinal
+	if arg.Name != "" {
+		debugName = arg.Name
+		index := s.stmt.BindParameterIndexSearch(arg.Name)
+		if index == 0 {
+			return &Error{
+				Code:  sqliteh.SQLITE_MISUSE,
+				Loc:   "Bind",
+				Query: s.query,
+				Msg:   fmt.Sprintf("unknown parameter name %q", arg.Name),
+			}
+		}
+		arg.Ordinal = index
+	}
+
+	// Start with obvious types, including time.Time before TextMarshaler.
+	found, err := s.bindBasic(debugName, arg.Ordinal, arg.Value)
+	if err != nil {
+		return err
+	} else if found {
+		return nil
+	}
+
+	if m, _ := arg.Value.(encoding.TextMarshaler); m != nil {
+		b, err := m.MarshalText()
+		if err != nil {
+			// TODO: modify Error to carry an error so we can %w?
+			return &Error{
+				Code:  sqliteh.SQLITE_MISUSE,
+				Loc:   "Bind",
+				Query: s.query,
+				Msg:   fmt.Sprintf("Bind:%v: cannot marshal %T: %v", debugName, arg.Value, err),
+			}
+		}
+		_, err = s.bindBasic(debugName, arg.Ordinal, b)
+		return err
+	}
+
+	// Look for named basic types or other convertible types.
+	val := reflect.ValueOf(arg.Value)
+	typ := reflect.TypeOf(arg.Value)
+	switch typ.Kind() {
+	case reflect.Bool:
+		b := int64(0)
+		if val.Bool() {
+			b = 1
+		}
+		_, err := s.bindBasic(debugName, arg.Ordinal, b)
+		return err
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		_, err := s.bindBasic(debugName, arg.Ordinal, val.Int())
+		return err
+	case reflect.Uint, reflect.Uint64:
+		return &Error{
+			Code:  sqliteh.SQLITE_MISUSE,
+			Loc:   "Bind",
+			Query: s.query,
+			Msg:   fmt.Sprintf("Bind:%v: sqlite does not support uint64", debugName),
+		}
+	case reflect.Uint8, reflect.Uint16, reflect.Uint32:
+		_, err := s.bindBasic(debugName, arg.Ordinal, int64(val.Uint()))
+		return err
+	case reflect.Float32, reflect.Float64:
+		_, err := s.bindBasic(debugName, arg.Ordinal, val.Float())
+		return err
+	case reflect.String:
+		_, err := s.bindBasic(debugName, arg.Ordinal, val.String())
```

Comment:
It seems silly to allocate to put val.String() into an interface{} and then promptly type switch on the type to execute a single line of code and a defer. Ditto for most of the other calls to bindBasic. I suggest exploding bindBasic into separate methods, each of which you can call. Or inline here by adding defer call above and calling `s.stmt.Bind*` directly here.

Reply:
I agree. I spent a bit of time trying to fix this without making a mess of the errors. I think it needs more tests before I can do it safely.